### ISSUE TITLE
Pagination

### ISSUE THREAD

**Issue by colbyfayock** [colbyfayock/next-wordpress-starter#9]
Determine a strategy for pagination or loading more results for a list of posts

## Questions
* Should it be "load more"?
* Should we use traditional pagination with numbers, next, and prev?
* Would the pagination be page-based? Would that use getStaticPaths?

Before this is picked up, please propose a plan for moving forward on how to go about this both technically and how it will appear in the UI

**Comment by colbyfayock** [comment]
We should probably wait to do this until the GraphQL Issue is resolved

https://github.com/colbyfayock/next-wordpress-starter/issues/5

**Comment by colbyfayock** [comment]
GraphQL is now used throughout the project so this is fair game

**Comment by doingandlearning** [comment]
I'm a big fan of *not* paginating and lazy loading the content if it's too much ... but for those who love the pagination happy to look at this.

**Comment by colbyfayock** [comment]
yeah a quick search seems to support that. as long as we can build it in where it's pretty clean between all devices

this seems to support it: https://www.nngroup.com/articles/infinite-scrolling/

id be curious to see if there are any more recent articles with more information on modern ways to support it and provide a good UX

**Comment by GuilleAngulo** [comment]
this one is interesting, from SEO perspective: https://developers.google.com/search/docs/guides/lazy-loading#paginated-infinite-scroll

**Comment by colbyfayock** [comment]
oooo thats really interesting

so i would imagine it would be along the lines of:
* Page loads with initial page 0 request
* Once you hit the trigger point, a new request is made for page 1
* Once page 1 is successfully requested, update the URL to something like ?page=1
* Do the same for page 2

When someone comes back with ?page1:
* Load page 0 and page 1 (making sure they load in the correct order)
* Update scroll position to page 1

The only other consideration i have with infinite scrolling is when implementing it with a site footer, it makes it very difficult for someone to interact with

it may make sense to only implement something like this on archive pages, such as all posts, categories, etc. then on the homepage, it would be a link to "view more" or w/e

**Comment by GuilleAngulo** [comment]
this one is a good read too, comparing [pagination - view more - infinite scroll:](https://uxdesign.cc/ui-cheat-sheet-pagination-infinite-scroll-and-the-load-more-button-e5c452e279a8) . 

I would say that classic pagination also fits in order to reference content. And navigation to older posts would be faster.

**Comment by colbyfayock** [comment]
maybe it makes sense to do pagination for now. it's a pretty classic way of doing things and would probably be a common expectation for someone spinning up a site (it's not controversial). i think the work that goes into it could be utilized to add lazy loading or something in the future should we go that direction since ultimately the paging needs to be requested

**Comment by GuilleAngulo** [comment]
After doing some research, here are some notes about that may be helpful:

WPgraphql supports an out-of-the-box cursor pagination (prev - next) based on [Relay connection specification](https://relay.dev/graphql/connections.htm). Apparently offset based pagination requires aditional installations ([for example](https://github.com/valu-digital/wp-graphql-offset-pagination)).

So when querying posts there are 4 variables available for this: `first` and `after` (post count) for forward pagination, `last` and `before` (cursors) for backward pagination. Also in the query it is possible to retrieve **pageInfo** (along with **edges**) from which it can be extracted `hasNextPage`, `hasPreviousPage`, `endCursor` and `startCursor` info. Here a [great article](https://www.wpgraphql.com/2020/03/26/forward-and-backward-pagination-with-wpgraphql/) on this, also with an explanation on page-based pagination downsides.

![pagination](https://user-images.githubusercontent.com/50624358/107696909-50684300-6c91-11eb-8902-30ef04d5551a.png)

I have made a test of this pagination for a better understanding following this flow: 

1. Inside **getServerSideProps** the **apolloClient** is initialized and it is made the query of first 10 posts. Then the cache of **apolloClient** is extracted and returned as a prop, `initialApolloState`. This prop is used in **useApolloClient** hook at `_app.js`, making the cache available .
```
export async function getStaticProps() {
 const apolloClient = initializeApollo()
 await apolloClient.query( QUERY_PAGINATED_POSTS, ... )
 return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
    },
  };
}
```

2. Inside the component the data is retrieve from cache using useQuery hook. It is picked up from the hook the **fetchMore** function, that allows to make another query in order to retrieve prev / next  posts (client side).

**In summary:** Index page will have 10 posts server side rendered within the HTML. The following posts using pagination will be client side rendered.

Not sure if this is the best approach, or if could be better try another thing. Would like to know your opinion.
There are also two problems related to this approach:

- How to deal with sticky posts. With this approach is not possible to have all posts info, then it occurs to me that we could only sort sticky posts inside each chunk. Would be a correct solution?
- After making the implementation I noticed there is a bug, and backward pagination info is coming swapped. Found that is an [open issue](https://github.com/wp-graphql/wp-graphql/issues/1552). 

Sorry about the long text, just wanted to put in order the ideas 😅

**Comment by colbyfayock** [comment]
so the only issue with doing something like that is we'll be then taking advantage of serverside rendering

up until this point i've been doing everything statically

if possible id like to continue that, i think its a nice challenge for solving those types of problems without requiring a server (and it makes it more portable)

as far as implementation, what do you think about using `getAllPosts` and use `getStaticPaths` to create the pages, and we can use that information to pass the available pagination into the tempalte to show the links to the pages

for instance https://next-wordpress-starter.netlify.app/posts/ has all posts (and this logic could apply to any "archive" page

inside `getStaticPaths` we break down how many posts divided by how many we want per page (maybe 10?), and create dynamic paths based on that. i think we may need to update the filename to make that work though, anyways...

then we have /posts/1, /posts/2, etc

i think the getAllPosts would be cached between building different pages right? so it wouldn't add a ton of overhead. i still need to see if i can confirm thats how that works 🤔

**Comment by GuilleAngulo** [comment]
Agree, that makes sense and it is in line with the whole project. Will try to make an implementation test to see if I can help on this one.

About **post per page**, could it be a constant (set by default to 10) placed at some "constants file"? In order to make it easy change that value?


> i think the getAllPosts would be cached between building different pages right? so it wouldn't add a ton of overhead.

Interesting point, will see if I can find some info about it also.

**Comment by colbyfayock** [comment]
yeah i think the constant with a default would be perfect. maybe that can go in next.config.js? unless you dont think thats appropriate. id love to have a central place, a constnat file makes sense, but wondering since we already have that file

**Comment by GuilleAngulo** [comment]
Yes, that seems like a good location along with the others env variables. 
I have one doubt about the implementation: where should the dynamic path be located? 
Because under `posts` directory is located the dynamic page for slugs `[slug].js`. Should pagination have a separated folder? Or should that dynamic page be shared for slugs and pages, having conditional blocks for each one?

**Comment by GuilleAngulo** [comment]
Yes, I think that is a good idea storing that constant along the other env variables. 

Sorry about the delay on this, but I have a question concerning the location of pagination: where should the dynamic paths be located at? Because under `posts` directory there is the `[slug].js` page ... should this page be in charge for both **pages** and **slugs**? Or should we create another directory, like for example `posts/page/2` -> `posts/page/[page].js` ? (not sure if I'm doing the correct questions)... can you give me a hand with this?

**Comment by colbyfayock** [comment]
yeah good question. i think in my opinion i'd like to see you go the page route like `posts/page/[page].js`

i'd prefer not to have the single Post page maintained along side the post pagination page and that's probably the cleanest way to handle it

do you think we could also use the `templates/archive.js` file to render out each of those pages? similar to the All Posts page:

https://github.com/colbyfayock/next-wordpress-starter/blob/master/src/pages/posts.js

also something to keep in imnd, we'll want to add relevant metadata to these pages to make sure that they're not hurting the site SEO wise

https://developers.google.com/search/blog/2011/09/pagination-with-relnext-and-relprev

i'm not sure if those pages additionally need a canonical tag pointing to the top level /all-posts page 🤔

**Comment by GuilleAngulo** [comment]
Took all the notes and I'm preparing the PR. Thanks a lot for the guidance!

About SEO:
I followed what you suggested. Aparently Google is no longer taking the **rel="next"** and **rel="prev"** as an indexing signal, but is still recommended by the [W3C](https://www.w3.org/TR/relations.html) and seems like Bing is also using it. So I have use them. And about _canonical_ what I have come up is only to specify the canonical for the fist page, where `/posts/` and `/posts/page/1` share the same content and both has the canonical link to `/posts/`. For the rest of the pages I have not added anything in that sense, because I thought that every page will have different posts. 
Another possibility (as I have seen in some articles) is to add a canonical for each page: Page 2 -> `<link rel="canonical" href=".../posts/page/2" />`. What do you think?

**Comment by colbyfayock** [comment]
yeah  idont feel strongly about the canonical link, whatever you think makes sense! and thats interesting about the tags, didn't realize that but good to know. for anyone tuning in https://ahrefs.com/blog/rel-prev-next-pagination/ is a good article to check out (just looked it up)
